File: js/1351.js
const fs = require('fs');

const [N, P, Q] = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split(' ')
  .map(Number);

const dic = new Map();

const ans = recur(N);
console.log(ans);

function recur(i) {
  if (i === 0) return 1;
  const ip = Math.floor(i / P);
  const iq = Math.floor(i / Q);

  if (dic[i]) return dic[i];
  else return (dic[i] = recur(ip) + recur(iq));
}

File: c++/1351.cpp
#include <iostream>
// #include <map>
#include <unordered_map>

using namespace std;

typedef long long ll;

ll N, P, Q;

unordered_map<ll, ll> dic;

ll recur(ll i) {
  if (i == 0) return 1;

  ll ip = i / P;
  ll iq = i / Q;

  if (dic[i])
    return dic[i];
  else
    return dic[i] = recur(ip) + recur(iq);
}

int main() {
  freopen("example.txt", "r", stdin);
  cin >> N >> P >> Q;

  cout << recur(N);

  return 0;
}